lib/posix-process/signal/signal.c

```diff
@@ -0,0 +1,267 @@
+	return sig;
+}
+
+struct uk_signal *pprocess_signal_next_pending_t(struct posix_thread *pthread)
+{
+	struct posix_process *pproc;
+	struct uk_signal *sig;
+
+	pproc = tid2pprocess(pthread->tid);
+	UK_ASSERT(pproc);
+
+	for (int i = 1; i < SIG_ARRAY_COUNT; i++) {
+		if (!pprocess_signal_is_deliverable(pthread, i))
+			continue;
+
+		sig = pprocess_signal_dequeue(pproc, pthread, i);
+		if (sig)
+			return sig;
+	}
+
+	return NULL;
+}
+
+void pprocess_signal_clear_pending(struct posix_process *proc, int signum)
+{
+	struct posix_thread *thread, *threadn;
+	struct uk_signal *sig;
+
+	if (IS_PENDING(proc->signal->sigqueue, signum)) {
+		while ((sig = pprocess_signal_dequeue(proc, NULL, signum)))
+			uk_signal_free(proc->_a, sig);
+	}
+
+	uk_pprocess_foreach_pthread(proc, thread, threadn) {
+		if (!IS_PENDING(thread->signal->sigqueue, signum))
+			continue;
+		while ((sig = pprocess_signal_dequeue(proc, thread, signum)))
+			uk_signal_free(thread->_a, sig);
+	}
+}
+
+void pprocess_signal_sigaction_clear(struct kern_sigaction *ks)
+{
+	ks->ks_handler = SIG_DFL;
+	ks->ks_flags = 0;
+	ks->ks_restorer = NULL;
+	uk_sigemptyset(&ks->ks_mask);
+}
+
+int pprocess_signal_sigaction_new(struct uk_alloc *alloc, struct uk_signal_pdesc *pd)
+{
+	pd->sigaction = uk_malloc(alloc, sizeof(struct uk_sigaction));
+	if (unlikely(!pd->sigaction)) {
+		uk_pr_err("Could not allocate memory\n");
+		return -ENOMEM;
+	}
+
+	pd->sigaction->alloc = alloc;
+	uk_refcount_init(&pd->sigaction->refcnt, 1);
+
+	return 0;
+}
+
+int pprocess_signal_pdesc_alloc(struct posix_process *process)
+{
+	UK_ASSERT(process);
+
+	process->signal = uk_malloc(process->_a,
+				    sizeof(struct uk_signal_pdesc));
+	if (unlikely(!process->signal)) {
+		uk_pr_err("Could not allocate memory\n");
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+int pprocess_signal_tdesc_alloc(struct posix_thread *thread)
+{
+	UK_ASSERT(thread);
+
+	thread->signal = uk_malloc(thread->_a, sizeof(struct uk_signal_tdesc));
+	if (unlikely(!thread->signal)) {
+		uk_pr_err("Could not allocate memory\n");
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+int pprocess_signal_pdesc_init(struct posix_process *process)
+{
+	struct uk_signal_pdesc *pd;
+	int rc;
+
+	UK_ASSERT(process);
+	UK_ASSERT(process->signal);
+
+	pd = process->signal;
+
+	pd->queued_count = 0;
+	pd->queued_max = _POSIX_SIGQUEUE_MAX;
+
+	uk_sigemptyset(&pd->sigqueue.pending);
+	for (__sz i = 0; i < SIG_ARRAY_COUNT; i++)
+		UK_INIT_LIST_HEAD(&pd->sigqueue.list_head[i]);
+
+	rc = pprocess_signal_sigaction_new(process->_a, pd);
+	if (unlikely(rc))
+		return rc;
+
+	for (__sz i = 0; i < SIG_ARRAY_COUNT; i++)
+		pprocess_signal_sigaction_clear(KERN_SIGACTION(process, i));
+
+	return 0;
+}
+
+int pprocess_signal_tdesc_init(struct posix_thread *thread)
+{
+	struct uk_signal_tdesc *td;
+
+	UK_ASSERT(thread);
+	UK_ASSERT(thread->signal);
+
+	td = thread->signal;
+
+	uk_sigemptyset(&td->mask);
+	uk_sigemptyset(&td->sigqueue.pending);
+
+	for (size_t i = 0; i < SIG_ARRAY_COUNT; i++)
```

Comment:
Is there a reason we are using `size_t` here but not in `pprocess_signal_pdesc_init` or the others?

Reply:
I don't think so, will change to `__sz`